src/WebJobs.Script/Host/AggregateFunctionMetadataProvider.cs

```diff
@@ -17,44 +17,69 @@
         {
             _logger = logger;
-            _dispatcher = invocationDispatcher ?? throw new ArgumentNullException(nameof(invocationDispatcher));
+            _dispatcher = invocationDispatcher;
+            _hostFunctionMetadataProvider = hostFunctionMetadataProvider;
         }
 
         public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors
            => _functionErrors.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.ToImmutableArray());
 
-        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, bool forceRefresh)
+        public async Task<ImmutableArray<FunctionMetadata>> GetFunctionMetadataAsync(IEnumerable<RpcWorkerConfig> workerConfigs, IEnvironment environment, bool forceRefresh)
         {
             IEnumerable<RawFunctionMetadata> rawFunctions = new List<RawFunctionMetadata>();
             IEnumerable<FunctionMetadata> functions = new List<FunctionMetadata>();
             _logger.FunctionMetadataProviderParsingFunctions();
+
+            bool workerIndexing = Utility.CanWorkerIndex(workerConfigs, environment);
+
             if (_functions.IsDefaultOrEmpty || forceRefresh)
             {
-                if (_dispatcher != null)
+                if (_dispatcher == null && workerIndexing == true)
+                {
+                    throw new ArgumentNullException(nameof(_dispatcher));
+                }
+                else if (_dispatcher != null && workerIndexing == true)
                 {
                     // start up GRPC channels
                     await _dispatcher.InitializeAsync(new List<FunctionMetadata>());
 
                     // get function metadata from worker, then validate it
                     rawFunctions = await _dispatcher.GetWorkerMetadata();
-                    functions = ValidateMetadata(rawFunctions);
 
-                    // set up invocation buffers and send load requests
-                    await _dispatcher.FinishInitialization(functions);
+                    if (rawFunctions != null && rawFunctions.Any())
+                    {
+                        functions = ValidateMetadata(rawFunctions);
+                        // set up invocation buffers and send load requests
+                        await _dispatcher.FinishInitialization(functions);
+                    }
+                    else
+                    {
+                        // If worker denies indexing then falling back to the host for Indexing
+                        _functions = _hostFunctionMetadataProvider.GetFunctionMetadataAsync(workerConfigs, SystemEnvironment.Instance, forceRefresh).GetAwaiter().GetResult();
+                    }
                 }
+                else if (workerIndexing == false)
+                {
+                    _functions = _hostFunctionMetadataProvider.GetFunctionMetadataAsync(workerConfigs, SystemEnvironment.Instance, forceRefresh).GetAwaiter().GetResult();
```

Comment:
nit: might be cleaner to separate this out into its own if statement  instead of if/else, and add a check for rawFunctions == null, so that way we only need to write this line of code once

Reply:
We should also `await` here, rather than doing `.GetAwaiter().GetResult()`